File: Loenn/entities/digitalPortal.lua
local utils = require("utils")
local drawableSprite = require("structs.drawable_sprite")

local digitalPortal = {}

digitalPortal.name = "MaggyHelper/DigitalPortal"
digitalPortal.depth = -100
digitalPortal.fieldInformation = {
    transportDelay = {
        minimumValue = 0
    }
}
digitalPortal.placements = {
    {
        name = "two_way",
        data = {
            destinationId = "",
            transportDelay = 0.5,
            isTwoWay = true
        }
    },
    {
        name = "one_way",
        data = {
            destinationId = "",
            transportDelay = 0.5,
            isTwoWay = false
        }
    }
}

function digitalPortal.sprite(room, entity)
    local sprite = drawableSprite.fromTexture("objects/digital_portal/inactive", entity)
    sprite:setPosition(entity.x, entity.y)
    sprite:setJustification(0.5, 1.0)
    sprite.color = {0.0, 0.8, 1.0, 0.8}
    return sprite
end

function digitalPortal.selection(room, entity)
    return utils.rectangle(entity.x - 16, entity.y - 48, 32, 48)
end

return digitalPortal

File: Ahorn/entities/DigitalPortal.jl
module DigitalPortal

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/DigitalPortal" DigitalPortal(x::Integer, y::Integer, destinationId::String="", isTwoWay::Bool=true, transportDelay::Number=0.5)

const placements = Ahorn.PlacementDict(
    "two_way" => Ahorn.EntityPlacement(DigitalPortal),
    "one_way" => Ahorn.EntityPlacement(DigitalPortal)
)

function Ahorn.selection(entity::DigitalPortal)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DigitalPortal, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end
